### PR TITLE
Rewrite threaded canvas drawing for improved performance

### PR DESCRIPTION
**Pull request by cmeyer**
Reviewers: this is a lengthy, complicated change and I recommend not trying to understand each change individually, nor do I recommend comparing _changes_ as a whole because the old version was buggy and more complicated. Instead, I recommend reading the [explanation about how canvas items are supposed to work,](https://github.com/nion-software/nionui-tool/blob/d9eb8fafe60515508b2b5cb7e279a4a6df24f2b4/launcher/DocumentWindow.cpp#L2358-L2385) then reviewing the associated methods and classes listed below to look for logic or thread safety issues:

- `PyCanvas::PyCanvas` (theory of operation) (`DocumentWindow.cpp`)
- `PyCanvas::~PyCanvas` (`DocumentWindow.cpp`)
- `PyCanvas::continuePaintingSection` (`DocumentWindow.cpp`)
- `PyCanvas::paintEvent` (`DocumentWindow.cpp`)
- `PyCanvas::setBinarySectionCommands` (`DocumentWindow.cpp`)
- `PyCanvas::removeSection` (`DocumentWindow.cpp`)
- `PyCanvasRenderTask` (entire class) (`DocumentWindow.cpp`)
- `CanvasSection` (entire class) (`DocumentWindow.cpp`)
- `Canvas_draw_binary` (`Application.cpp`)
- `Canvas_drawSection_binary` (`Application.cpp`)

**NOTE** I've been doing my testing using the new drawing framework. Late tonight I tested with the main branch and there is an issue with initial canvas sizing. You can temporarily work around this by clicking in different panels to trigger redraw. I'm aware of the issue and will investigate tomorrow. Until then, the code still needs review and feedback.

---

- **Change repaint to run periodically rather than at request for improved performance.**
- **Rename 'rendering' to 'm_is_rendering'.**
- **Rework threaded painting to simplify and for performance.**
- **Fix Xcode issues.**
- **Use shared pointer for immutable drawing commands to avoid unnecessary copies.**
- **Fix issue with repaint queue.**
- **Dispatch task outside of lock for improved performance.**
- **Reconsolidate timer to fix repaint slowdown.**
- **Fix latency display, add low level frame rate display.**
- **Update documentation and names.**
- **Ensure commands, rect, and image map get updated as one. Add more documentation.**

### CODE REVIEW

**Review by Tiomat85 — APPROVED** [review]

**Review by Tiomat85 — CHANGES_REQUESTED** [review]

File: launcher/DocumentWindow.cpp [hunk 1]
```diff
@@ -2850,42 +2855,50 @@ void PyCanvas::setBinarySectionCommands(int section_id, const QSharedPointer<std
...
-    queueTask(section);
+        task = queueTask(section);
+    }
+
+    if (task)
+        QThreadPool::globalInstance()->start(task);
 }
 
-void PyCanvas::queueTask(QSharedPointer<CanvasSection> section)
+PyCanvasRenderTask *PyCanvas::queueTask(QSharedPointer<CanvasSection> section)
```
[the remark this answers is not in the document]
Comment (cmeyer) [reply launcher/DocumentWindow.cpp]: See comments in https://github.com/nion-software/nionui-tool/commit/642c5e437735d94dd7ef337cf12794493f286267

File: launcher/DocumentWindow.cpp [hunk 2]
```diff
@@ -2850,42 +2855,50 @@ void PyCanvas::setBinarySectionCommands(int section_id, const QSharedPointer<std
...
+        imageMapCopy = section->m_imageMap;
+        section->rect = rect;
+        section->m_imageMap = imageMap;
 
-    queueTask(section);
+        task = queueTask(section);
+    }
+
+    if (task)
+        QThreadPool::globalInstance()->start(task);
```
[the remark this answers is not in the document]
Comment (cmeyer) [reply launcher/DocumentWindow.cpp:2973]: See comments in https://github.com/nion-software/nionui-tool/commit/642c5e437735d94dd7ef337cf12794493f286267

File: launcher/DocumentWindow.cpp [hunk 3]
```diff
@@ -2356,13 +2310,36 @@ PyCanvasRenderTask::PyCanvasRenderTask(PyCanvas *canvas, const QSharedPointer<Ca
...
-
 {
 }
 
 void PyCanvasRenderTask::run()
 {
-    auto render_result = render(m_section, m_commands_binary, m_rect, m_image_map, m_device_pixel_ratio, m_rendered_timestamps);
+    RenderResult render_result(m_section);
+
+    if (m_commands_binary && !m_rect.isEmpty())
```
[the remark this answers is not in the document]
Comment (cmeyer) [reply launcher/DocumentWindow.cpp]: Yes, same as the other comment - technically wrong, but practically doesn't matter. I'll fix it to read more clearly.

File: launcher/DocumentWindow.h [hunk 4]
```diff
@@ -559,31 +561,21 @@ struct RenderResult
...
 public:
-    PyCanvasRenderTask(const QSharedPointer<CanvasSection> &section, const std::vector<quint32> &commands_binary, const QRect &rect, const QMap<QString, QVariant> &imageMap, float devicePixelRatio, const RenderedTimeStamps &rendered_timestamps);
-
-    PyCanvasRenderTaskSignals *signals() const { return m_signals.get(); }
+    PyCanvasRenderTask(PyCanvas *canvas, const QSharedPointer<CanvasSection> &section, const QSharedPointer<std::vector<quint32>> &commands_binary, const QRect &rect, const QMap<QString, QVariant> &imageMap, float devicePixelRatio, const RenderedTimeStamps &rendered_timestamps);
 
     virtual void run() override;
 
 private:
+    PyCanvas *m_canvas;
```
[the remark this answers is not in the document]
Comment (Tiomat85) [reply launcher/DocumentWindow.h:598]: Ok, if you are happy about lifetimes of the parent creater object being longer than this then that is fine. A reference might be better, but we are talking failure cases at that point.

**Pull request merged**